pkg/netpol/scan/scan.go

```diff
@@ -109,43 +134,112 @@ const (
+// return if yamlDoc is of accepted kind, and the kind string
+func getKind(yamlDoc string) (bool, string) {
+	if yamlDoc == "\n" || yamlDoc == "" {
+		// ignore empty cases
+		return false, ""
+	}
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	_, groupVersionKind, err := decode([]byte(yamlDoc), nil, nil)
+	if err != nil {
+		return false, ""
+	}
+	if !acceptedK8sTypes.MatchString(groupVersionKind.Kind) {
+		fmt.Printf("Skipping object with type: %s", groupVersionKind.Kind)
+		return false, ""
+	}
+	return true, groupVersionKind.Kind
 }
 
-func GetYAMLDocumentsFromPath(repoDir string, walkFn WalkFunction) []string {
-	res := make([]string, 0)
-	manifestFiles := searchDeploymentManifests(repoDir, walkFn)
-	for _, mfp := range manifestFiles {
-		filebuf, err := os.ReadFile(mfp)
-		if err != nil {
-			continue
-		}
-		res = append(res, splitByYamlDocuments(filebuf)...)
+// given a YAML doc and its resource kind, convert to a slice of K8sObject
+func mainfestToK8sObjects(yamlDoc, kind string) ([]K8sObject, error) {
+	objDataBuf := []byte(yamlDoc)
+	res := make([]K8sObject, 0, 1)
+	switch kind {
+	case Pod:
+		obj := parsePod(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{Pod: obj, Kind: kind})
+	case ReplicaSet:
+		obj := parseReplicaSet(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{Replicaset: obj, Kind: kind})
+	case Deployment:
+		obj := parseDeployment(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{Deployment: obj, Kind: kind})
+	case Statefulset:
+		obj := parseStatefulSet(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{Statefulset: obj, Kind: kind})
+	case ReplicationController:
+		obj := parseReplicationController(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{ReplicationController: obj, Kind: kind})
+	case Daemonset:
+		obj := parseDaemonSet(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{Daemonset: obj, Kind: kind})
+	case Job:
+		obj := parseJob(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{Job: obj, Kind: kind})
+	case CronJob:
+		obj := parseCronJob(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{CronJob: obj, Kind: kind})
+	case Networkpolicy:
+		obj := parseNetworkPolicy(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{Networkpolicy: obj, Kind: kind})
+	case Namespace:
+		obj := parseNamespace(bytes.NewReader(objDataBuf))
+		res = append(res, K8sObject{Namespace: obj, Kind: kind})
+	case List:
+		res = parseList(objDataBuf)
+	case PodList:
+		res = parsePodList(objDataBuf)
+	case NamespaceList:
+		res = parseNamespaceList(objDataBuf)
+	// TODO: support other specific list types, other than PodList and NamespaceList
+	default:
+		return res, fmt.Errorf("unsupported kind: %s", kind)
 	}
-	return res
+	return res, nil
 }
 
+const yamlParseBufferSize = 200
+
 // return a list of paths for yaml files in the input dir path
 func searchDeploymentManifests(repoDir string, walkFn WalkFunction) []string {
```

Comment:
Function name is misleading - it's not only deployments this function searches for.
Better call it `searchYamlFiles`.

Reply:
fixed.